File: xwzx-news/src/store/modules/notification.js
import { defineStore } from 'pinia'
import axios from 'axios'

import { apiConfig } from '../../config/api'
import { useUserStore } from '../user'


const requestError = (error, fallback) => ({
  success: false,
  status: error.response?.status,
  message: error.response?.data?.message || fallback,
})

export const useNotificationStore = defineStore('notification', {
  state: () => ({
    notifications: [],
    page: 1,
    pageSize: 20,
    total: 0,
    hasMore: false,
    unreadCount: 0,
    loading: false,
  }),

  actions: {
    authHeaders() {
      const userStore = useUserStore()
      return { Authorization: `Bearer ${userStore.token}` }
    },

    async fetchUnreadCount() {
      try {
        const response = await axios.get(
          `${apiConfig.baseURL}/api/notifications/unread-count`,
          { headers: this.authHeaders() },
        )
        this.unreadCount = response.data?.data?.unreadCount || 0
        return { success: true, unreadCount: this.unreadCount }
      } catch (error) {
        return requestError(error, '获取未读通知数量失败')
      }
    },

    async fetchNotifications(page = 1) {
      this.loading = true
      try {
        const response = await axios.get(`${apiConfig.baseURL}/api/notifications`, {
          headers: this.authHeaders(),
          params: { page, pageSize: this.pageSize },
        })
        const data = response.data?.data || {}
        const items = data.list || []
        this.notifications = page === 1 ? items : [...this.notifications, ...items]
        this.page = page
        this.total = data.total || 0
        this.hasMore = Boolean(data.hasMore)
        this.unreadCount = data.unreadCount || 0
        return { success: true }
      } catch (error) {
        return requestError(error, '获取通知列表失败')
      } finally {
        this.loading = false
      }
    },

    async markRead(notificationId) {
      const item = this.notifications.find((notice) => notice.id === notificationId)
      if (item?.isRead) return { success: true }
      try {
        await axios.patch(
          `${apiConfig.baseURL}/api/notifications/${notificationId}/read`,
          {},
          { headers: this.authHeaders() },
        )
        if (item) item.isRead = true
        this.unreadCount = Math.max(0, this.unreadCount - 1)
        return { success: true }
      } catch (error) {
        return requestError(error, '标记通知已读失败')
      }
    },

    async markAllRead() {
      try {
        await axios.patch(
          `${apiConfig.baseURL}/api/notifications/read-all`,
          {},
          { headers: this.authHeaders() },
        )
        this.notifications.forEach((item) => { item.isRead = true })
        this.unreadCount = 0
        return { success: true }
      } catch (error) {
        return requestError(error, '全部标记已读失败')
      }
    },

    async deleteNotification(notificationId) {
      try {
        await axios.delete(`${apiConfig.baseURL}/api/notifications/${notificationId}`, {
          headers: this.authHeaders(),
        })
        const item = this.notifications.find((notice) => notice.id === notificationId)
        if (item && !item.isRead) this.unreadCount = Math.max(0, this.unreadCount - 1)
        this.notifications = this.notifications.filter((notice) => notice.id !== notificationId)
        this.total = Math.max(0, this.total - 1)
        return { success: true }
      } catch (error) {
        return requestError(error, '删除通知失败')
      }
    },

    async clearNotifications() {
      try {
        await axios.delete(`${apiConfig.baseURL}/api/notifications`, {
          headers: this.authHeaders(),
        })
        this.notifications = []
        this.total = 0
        this.hasMore = false
        this.unreadCount = 0
        return { success: true }
      } catch (error) {
        return requestError(error, '清空通知失败')
      }
    },

    async publishAnnouncement(payload) {
      try {
        const response = await axios.post(
          `${apiConfig.baseURL}/api/admin/notifications/announcements`,
          payload,
          { headers: this.authHeaders() },
        )
        return { success: true, data: response.data?.data }
      } catch (error) {
        return requestError(error, '发布公告失败')
      }
    },
  },
})

File: routers/notification.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from config.db_conf import get_db
from crud import notification
from models.users import User
from schemas.notification import (
    NotificationItemResponse,
    NotificationListResponse,
    UnreadCountResponse,
)
from utils.auth import get_current_user
from utils.response import success_response


router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get("")
async def get_notifications(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100, alias="pageSize"),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    rows, total, unread_count = await notification.get_notification_list(
        db, user.id, page, page_size
    )
    data = NotificationListResponse(
        list=[NotificationItemResponse.model_validate(item) for item in rows],
        total=total,
        hasMore=page * page_size < total,
        unreadCount=unread_count,
    )
    return success_response(message="获取通知列表成功", data=data)


@router.get("/unread-count")
async def get_unread_count(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    count = await notification.get_unread_count(db, user.id)
    return success_response(
        message="获取未读数量成功",
        data=UnreadCountResponse(unreadCount=count),
    )


@router.patch("/read-all")
async def mark_all_read(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    count = await notification.mark_all_notifications_read(db, user.id)
    return success_response(message=f"已将{count}条通知标记为已读")


@router.patch("/{notification_id}/read")
async def mark_read(
    notification_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    item = await notification.mark_notification_read(db, user.id, notification_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="通知不存在")
    return success_response(
        message="通知已读", data=NotificationItemResponse.model_validate(item)
    )


@router.delete("/{notification_id}")
async def remove_notification(
    notification_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    deleted = await notification.delete_notification(db, user.id, notification_id)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="通知不存在")
    return success_response(message="删除通知成功")


@router.delete("")
async def clear_notifications(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    count = await notification.clear_notifications(db, user.id)
    return success_response(message=f"清空了{count}条通知")

